Parse/src/main/java/com/parse/NetworkQueryController.java

```diff
@@ -134,9 +134,12 @@ public Integer then(Task<JSONObject> task) throws Exception {
       if (resultClassName == null) {
         resultClassName = state.className();
       }
+      boolean isSubset = state.selectedKeys() != null;
+      JSONArray selectedKeys = isSubset ? new JSONArray(state.selectedKeys()) : null;
       for (int i = 0; i < results.length(); ++i) {
         JSONObject data = results.getJSONObject(i);
-        T object = ParseObject.fromJSON(data, resultClassName, state.selectedKeys() == null);
+        if (isSubset) data.put(ParseObject.KEY_SELECTED_KEYS, selectedKeys);
```

Comment:
@rogerhu I get your point, I tried working directly with a Set, but for that we would need to change the signatures of ParseObject.fromJSON, ParseObject.mergeFromServer, and even worse ParseDecoder.decode, these are called from anywhere and an extra set of "selected keys" rarely makes sense.
The key is also used by ParseDecoder to get keys from the JSON...

Maybe we could add two static methods to ParseObject, `void insertSelectedKeys(JSONObject data, Set<String>)` , used by NetworkQueryController, and `boolean hasSelectedKeys(JSONObject data)` , used by ParseDecoder. What do you think?